Use timeslot type to identify office hours sessions for agenda filtering and hide jabber (etc) links for office hours sessions

jennifer-richards opened the pull request:
This is the last piece of the fix for [ticket 3115](https://trac.ietf.org/trac/ietfdb/ticket/3115) and [ticket 3128](https://trac.ietf.org/trac/ietfdb/ticket/3128). This builds on the work from #10 which is still waiting for sign-off, so it's probably best to wait until that is approved before digging in to this one.

This finally switches the agenda filtering system from identifying timeslots as office hours by the string appearing in the session name to labeling them with an 'officehours' timeslot type. A benefit of this is that the office hours can now belong to the area group to which they apply instead all being assigned to the IESG.

For "legacy" meetings (i.e., those prior to 111), the old method is kept. To manage the complexity this introduces, the `AgendaKeywordTagger` class is introduced as a sibling class of the `AgendaFilterOrganizer` class from the previous pull request's work. These are children of a common base class that contains common logic - mostly the definition of which meetings constitute 'legacy' meetings.

This finally allows the issue described in ticket 3128 (jabber links and other agenda buttons appearing for office hours sessions, to which they do not apply) to be resolved.

This work was done somewhat out of order and has been rebased a couple of times, so please keep an eye out for merge accidents. I found a couple places where PyCharm misunderstood patches and doubled code - it's possible there are more, though none that the tests catch.

meadmaker commented:
Just for the fun of it, we've now had another meeting, and another couple/few releases.  I imagine this will take another round of rebasing and merging, as well as updates to the "legacy" meeting definition.

I'm curious why you decided to code in two different modes of operation - the "legacy" meetings and the current way to do things - rather than deciding to code one way to operate and migrating the legacy data.  Would that have been more difficult?

jennifer-richards commented:
I discussed that with Robert before implementing this. He was concerned that the migrations would risk urgent fixes if there were problems and preferred keeping legacy handling. (He mentioned, e.g., worry about unintended consequences affecting things like proceedings.)

meadmaker reviewed (changes requested):
There are a couple of places where changes are needed, and others that are more questions rather than demands for changes.

jennifer-richards commented on ietf/doc/templatetags/ietf_filters.py:
```diff
@@ -613,3 +613,80 @@ def action_holder_badge(action_holder):
+    >>> any(is_regular_agenda_item(factory(t)) for t in ['plenary', 'break', 'reg', 'other', 'officehours'])
+    False
+    """
+    return assignment.timeslot.type.slug == 'regular'
```
Good question.

This strikes me as a bit of a corner case for Demeter because the explicit purpose of the awkwardly named `SchedTimeSessAssignment` model is to link a session with a timeslot. Though not explicitly a method, `timeslot` is part of the public interface of the class. Python allows us to replace it with a `@property` getter method returning a `TimeSlot` if we need to change the inner workings.

On the other hand, I find it a bit arbitrary that the `TimeSlot` and not the `Session` defines the purpose of the assignment. It's not hard to imagine the rules for an assignment becoming more complicated. That, I think, argues for a method on `SchedTimeSessAssignment` being responsible for determining the type.

Naming, as usual, becomes a problem. Calling it `slot_type` is logical but the name suggests that it's just a pass-through to the timeslot. Calling it `type` would get in the way if there are ever different types of assignments (like "tentative" vs "fixed", e.g.). I'm not sure how to describe "the type of event represented by this assignment" succinctly. On the other other hand, it's returning a `TimeSlotTypeName`, so the relationship to a slot type is pretty well baked in.

Ultimately I agree it should be a method on the assignment, and I think calling it `slot_type()` is as sensible as anything.

Similarly, I will add a `meeting()` method to the class. I'm going to let references to `assignment.session.name` stand because I think the `Session` should be in charge of its name.

jennifer-richards commented on ietf/group/migrations/0043_groupfeatures_agenda_filter_type.py:
```diff
@@ -0,0 +1,42 @@
+# Generated by Django 2.2.19 on 2021-04-02 12:03
```
Will do

jennifer-richards commented on ietf/name/migrations/0024_agendafiltertypename.py:
```diff
@@ -0,0 +1,27 @@
+# Generated by Django 2.2.19 on 2021-04-02 12:03
```
Will do

jennifer-richards commented on ietf/group/migrations/0043_groupfeatures_agenda_filter_type.py:
```diff
@@ -0,0 +1,42 @@
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('name', '0024_agendafiltertypename'),
```
Yes it should (I think this was from an iteration where the schema and data migrations were in a single migration)

jennifer-richards commented on ietf/meeting/helpers.py:
```diff
@@ -243,39 +244,412 @@ def preprocess_assignments_for_agenda(assignments_queryset, meeting, extra_prefe
+
+        self.meeting = self.sessions[0].meeting if len(self.sessions) > 0 else None
+
+        self.filterable_timeslot_types = TimeSlotTypeName.objects.filter(
+            slug__in=self.filterable_timeslot_slugs
+        )
```
I think I was concerned it would cause a DB query when the module was imported. It may be ok because Django deals with queries lazily, I will check.

jennifer-richards replied:
I believe it's ok outside of init, so moving it.

jennifer-richards commented on ietf/meeting/helpers.py:
```diff
@@ -243,39 +244,412 @@ def preprocess_assignments_for_agenda(assignments_queryset, meeting, extra_prefe
+        super(AgendaFilterOrganizer, self).__init__(**kwargs)
+        self.single_category = single_category
+
+        # group acronyms in this list will never be used as filter buttons
+        self.exclude_acronyms = ['iesg', 'ietf', 'secretariat']
+
+        # extra keywords to include in the no-heading column if they apply to any sessions
+        self.extra_labels = ['BoF', 'Plenary']
+
+        # group types whose acronyms should be word-capitalized
+        self.capitalized_group_types = ['team']
+
+        # group types whose acronyms should be all-caps
```
Sounds good

jennifer-richards commented on ietf/meeting/views.py:
```diff
@@ -3561,7 +3441,7 @@ def proceedings(request, num=None):
     ietf      = sessions.filter(group__parent__type__slug = 'area').exclude(group__acronym='edu')
     irtf      = sessions.filter(group__parent__acronym = 'irtf')
-    training  = sessions.filter(group__acronym__in=['edu','iaoc'], type_id__in=['regular', 'other', ]).exclude(current_status='notmeet')
+    training  = sessions.filter(group__acronym__in=['edu','iaoc'], type_id__in=['regular', 'other', 'officehours',]).exclude(current_status='notmeet')
```
did not discuss, asking now

jennifer-richards replied:
As you probably saw, the answer is it should not (and this area will need to change soon anyway). Thanks for the catch.

jennifer-richards commented on ietf/doc/templatetags/ietf_filters.py:
```diff
@@ -613,3 +613,80 @@ def action_holder_badge(action_holder):
+    True
+    """
+    num = assignment.session.meeting.number
+    if num.isdigit() and int(num) <= 110:
```
Done (moving to a settings variable)

jennifer-richards commented on ietf/meeting/helpers.py:
```diff
@@ -243,39 +244,412 @@ def preprocess_assignments_for_agenda(assignments_queryset, meeting, extra_prefe
+        # by having a purely numeric meeting.number.
+        return (self.meeting is not None
+                and self.meeting.number.isdigit()
+                and int(self.meeting.number) <= 110)
```
Done

jennifer-richards commented:
I believe I've updated everything except the migration renumbering. I think that will be better handled along with merging (or perhaps rebasing) to the new release.

jennifer-richards commented:
I have rebased and updated for 7.37.1 - it was going to be a huge pain to merge the new upstream branches in. The result of that rebase plus some minor work to reconcile this branch with upstream developments are in jennifer/tix3128-rebase-7.37.1

meadmaker commented on ietf/templates/meeting/room-view.html:
```diff
@@ -24,7 +24,7 @@
      if (room_names.indexOf("{{ss.timeslot.get_hidden_location}}") >= 0 )
      {
-         items.push({room_index:room_names.indexOf("{{ss.timeslot.get_hidden_location}}"),day:{{ss.day}}, delta_from_beginning:{{ss.delta_from_beginning}},time:"{{ss.timeslot.time|date:"Hi"}}-{{ss.timeslot.end_time|date:"Hi"}}", verbose_time:"{{ss.timeslot.time|date:"D M d Hi"}}-{{ss.timeslot.end_time|date:"Hi"}}",duration:{{ss.timeslot.duration.total_seconds}}, type:"{{ss.timeslot.type}}", {% if ...
+         items.push({room_index:room_names.indexOf("{{ss.timeslot.get_hidden_location}}"),day:{{ss.day}}, delta_from_beginning:{{ss.delta_from_beginning}},time:"{{ss.timeslot.time|date:"Hi"}}-{{ss.timeslot.end_time|date:"Hi"}}", verbose_time:"{{ss.timeslot.time|date:"D M d Hi"}}-{{ss.timeslot.end_time|date:"Hi"}}",duration:{{ss.timeslot.duration.total_seconds}}, type:"{{ss.slot_type}}", {% if ss. ...
```
This took some real work to find the changes to review!

jennifer-richards replied:
Sorry. I was afraid to touch it any more than necessary...

meadmaker replied:
That's quite a good thought.  It wasn't a complaint, just... an expression of how I was feeling.  I wound up adding a newline before every closing brace, which narrowed down the changes pretty obviously.

jennifer-richards replied:
I think I did something similar when trying to figure out what to change, but I have apparently blocked most if it out of my memory.

I'm reminded that part of why I didn't do more to this was because it really ought to be restructured. Generating code via template is not a good practice, and really we should be trying to separate javascript out of the HTML files entirely.

But maybe I'll do a separate SVN commit just to break up this 1200 character line...

meadmaker commented on ietf/settings.py:
```diff
@@ -947,6 +947,7 @@ def skip_unreadable_post(record):
 FLOORPLAN_DIR = os.path.join(MEDIA_ROOT, FLOORPLAN_MEDIA_DIR)
 
 MEETING_USES_CODIMD_DATE = datetime.date(2020,7,6)
+MEETING_LEGACY_OFFICE_HOURS_END = 111  # last meeting to use legacy office hours representation
```
1:11 is a very odd time for the office hours to end...

jennifer-richards replied:
True

meadmaker reviewed (changes requested):
Back at you.  I've been doing this long enough that I can't remember whether I actually requested changes at any point along the way, but I assume that I did.

jennifer-richards commented on ietf/doc/templatetags/ietf_filters.py:
```diff
@@ -613,3 +613,89 @@ def action_holder_badge(action_holder):
+    return assignment.slot_type().slug == 'regular'
+
+@register.filter
+def is_plenary_agenda_item(assignment):
+    """Is this agenda item a regular session item?
+    
+    A regular item appears as a sub-entry in a timeslot within the agenda
```
Correct, and corrected.

jennifer-richards commented on ietf/doc/templatetags/ietf_filters.py:
```diff
@@ -613,3 +613,89 @@ def action_holder_badge(action_holder):
+    ]
+
+@register.filter
+def should_show_agenda_session_buttons(assignment):
+    """Should this agenda item show the session buttons (jabber link, etc)?
```
I've renamed it to `should_suppress_agenda_session_buttons`. I would like to avoid tying its name to its current logic. I've added the comments to clarify that it's for use in the `session_buttons_include.html` template.

Also, I'm moving this and the various `is_*_agenda_item` out of the `ietf/doc/templatetags/ietf_filters.py` file and into `ietf/meeting/templatetags/agenda_custom.py`. These methods have nothing to do with docs (I had assumed `ietf_filters.py` was a utils module, hadn't noticed it was in doc until recently).

meadmaker replied:
Oh, I hadn't even notice that directory discrepancy!  Good catch.

jennifer-richards commented on ietf/meeting/helpers.py:
```diff
@@ -243,39 +244,401 @@ def preprocess_assignments_for_agenda(assignments_queryset, meeting, extra_prefe
+        # n.b., single star argument means only keyword parameters are allowed when calling constructor
+        if assignments is not None and sessions is None:
+            self.assignments = assignments
+            self.sessions = [a.session for a in self.assignments if a.session]
```
I think it will bomb out with a `ValueError` or `KeyError` if it gets a scalar for either assignments or sessions. That seems pretty close to the right thing to do. It should point at the right place in the traceback, so I don't know that catching the error just to improve its error message adds a whole lot of value.

I can add type hints to help the linter understand out that these need to be iterables.

meadmaker replied:
Whatever you think is best - you're much more the expert than I.

jennifer-richards commented on ietf/meeting/helpers.py:
```diff
@@ -243,39 +244,401 @@ def preprocess_assignments_for_agenda(assignments_queryset, meeting, extra_prefe
+
+        These are the keywords corresponding to the right-most, headingless button column.
+        """
+        if self.special_filters is None:
+            self._organize_filters()
```
Premature optimization. I was worried that organizing the filters would be expensive. As it turns out, it's not too bad, and we never have a use case that involves instantiating the object and not immediately calling one of the methods that calls `_organize_filters`. I'll put it in `__init__()`

jennifer-richards commented:
I've pushed patches to the rebased branch (3a802211fd272e3e79b70c60e40de92bb3a361b1). I expect they'll apply here without effort, in which case I'll cherry pick them so you can compare

jennifer-richards commented:
This is going to be merged into 7.40.0 by way of ticket 3330. Closing without merge.
